doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,468 @@
+names: Integral, Floating, ComplexFloatin.
+
+When promotion occurs (as it does for ufuncs if no exact loop matches),
+the other DType is able to decide how to regard the Python
+scalar.  E.g. a ``UInt16`` promoting with an `Integral` will give ``UInt16``.
+
+.. note::
+
+    A default will most likely be provided in the future for user defined DTypes.
+    Most likely this will end up being the default integer/float, but in principle
+    more complex schemes could be implemented.
+
+At no time is the value used to decide the result of this promotion.  The value is only
+considered when it is converted to the new dtype; this may raise an error.
+
+
+
+
+Related Work
+============
+
+* `JAX promotion`_ also uses the weak-scalar concept.  However, it makes use
+  of it also for most functions.  JAX further stores the "weak-type" information
+  on the array: ``jnp.array(1)`` remains weakly typed.
+
+
+Implementation
+==============
+
+Implemeting this NEP requires some additional machinery to be added to all
+binary operators (or ufuncs), so that they attempt to use the "weak" logic
+if possible.
+There are two possible approaches to this:
+
+1. The binary operator simply tries to call ``np.result_type()`` if this
+   situation arises and converts the Python scalar to the result-type (if
+   defined).
+2. The binary operator indicates that an input was a Python scalar, and the
+   ufunc dispatching/promotion machinery is used for the rest (see
+   :ref:`NEP 42 <NEP42>`).  This allows more flexibility, but requires some
+   additional logic in the ufunc machinery.
+
+.. note::
+   As of now, it is not quite clear which approach is better, either will
+   give fairl equivalent results and 1. could be extended by 2. in the future
+   if necessary.
+
+It further requires removing all current special value-based code paths.
+
+Unintuitively, a larger step in the implementation may be to implement a
+solution to allow an error to be raised in the following example::
+
+   np.arange(10, dtype=np.uint8) + 1000
+
+Even though ``np.uint8(1000)`` returns the same value as ``np.uint8(232)``.
+
+.. note::
+
+    See alternatives, we may yet decide that this silent overflow is acceptable
+    or at least a separate issue.
+
+
+Alternatives
+============
+
+There are several design axes where different choices are possible.
+The below sections outline these.
+
+Use strongly typed scalars or a mix of both
+-------------------------------------------
+
+The simplest solution to the value-based promotion/casting issue would be to use
+strongly typed Python scalars, i.e. Python floats are considered double precision
+and Python integers are always considered the same as the default integer dtype.
+
+This would be the simplest solution, however, it would lead to many upcasts when
+working with arrays of ``float32`` or ``int16``, etc.  The solution for these cases
+would be to rely on in-place operations.
```

Comment:
It may be good to write it as an explicit advantage of the proposed scheme that operations with a python scalar give the same result whether or not it is done in-place.